src/sage/homology/homology_vector_space_with_basis.py

```diff
@@ -385,6 +393,67 @@ def _to_cycle_on_basis(self, i):
                                          cochains=self._cohomology)
         return chains.from_vector(vec)
 
+    def dual(self):
+        r"""
+        Return the dual space.
+
+        If ``self`` is homology, return the cohomology ring. If
+        ``self`` is cohomology, return the homology as a vector space.
+
+        EXAMPLES::
+
+            sage: T = simplicial_complexes.Torus()
+            sage: hom = T.homology_with_basis(GF(2))
+            sage: coh = T.cohomology_ring(GF(2))
+            sage: hom.dual() is coh
+            True
+            sage: coh.dual() is hom
+            True
+        """
+        if self.base_ring() == GF(2):
+            if self._cohomology:
+                return HomologyVectorSpaceWithBasis_mod2(self.base_ring(),
+                                                         self.complex())
+            return CohomologyRing_mod2(self.base_ring(), self.complex())
+        if self._cohomology:
+            return HomologyVectorSpaceWithBasis(self.base_ring(),
+                                                self.complex(),
+                                                not self._cohomology)
+        return CohomologyRing(self.base_ring(), self.complex())
+
+    def _test_duality(self, **options):
+        r"""
+        Test if the ordered bases for homology and cohomology are compatible.
+        Return nothing if the test succeeds.
+
+        This checks whether each evaluation map `H^n \otimes H_n \to
+        k` is represented by the identity matrix, in terms of the
+        chosen bases.
+
+        TESTS::
+
+            sage: T = simplicial_complexes.Torus()
+            sage: K = T.suspension()
+            sage: K.set_immutable()
+            sage: H = K.cohomology_ring(QQ)
+            sage: H._test_duality()
+
+            sage: simplicial_complexes.RandomComplex(8, 2, .2).homology_with_basis(GF(2))._test_duality()
+            sage: simplicial_complexes.RandomComplex(8, 2, .4).homology_with_basis(GF(2))._test_duality()
+            sage: simplicial_complexes.RandomComplex(8, 2, .6).homology_with_basis(GF(2))._test_duality()
+
+            sage: simplicial_complexes.RandomComplex(12, 3, .5).homology_with_basis(GF(2))._test_duality() # long time
+        """
+        okay = True
+        dual = self.dual()
+        dims = [a[0] for a in self._indices]
+        for dim in range(max(dims) + 1):
+            n = len(self.basis(dim))
+            m = matrix(n, n, [a.eval(b) for a in self.basis(dim) for b in dual.basis(dim)])
+            okay = (m == 1)
+            if not okay:
+                print('error in dimension {}'.format(dim))
```

Comment:
```suggestion
        tester = self._tester(**options))
        dual = self.dual()
        dims = [a[0] for a in self._indices]
        for dim in range(max(max(dims),  tester._max_runs) + 1):
            n = len(self.basis(dim))
            m = matrix(n, n, [a.eval(b) for a in self.basis(dim) for b in dual.basis(dim)])
            tester.assertEqual(m, 1,f "error in dimension {dim}")
```
This is the standard `_test_*` method paradigm. This also allows you to limit the dimensions checked.